### PR TITLE
feat: integrate wallchain

### PR DESCRIPTION
**Pull request by fish-sausage**
Integration with [Wallchain](https://www.wallchain.xyz/) to capture MEV.

### CODE REVIEW

**Comment by chef-jojo** [comment]
checklist: 
- we will only allow 1 pair for the first integration
- compare the original trade 
- sdk including the whole ethers v6, we might only need the api, sign permit, and create tx part.

**Comment by fish-sausage** [comment]
Addressed both issues.

File: apps/web/src/views/Swap/V3Swap/hooks/useWallchain.ts [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+    const sdk = new WallchainSDK({
+      keys: WallchainKeys,
+      provider: walletClient.transport as TOptions['provider'],
+      overrideAddresses,
+    })
+
+    if (
+      swapCalls[0] &&
+      swapCalls[0].calldata.toLowerCase().includes('190b589cf9Fb8DDEabBFeae36a813FFb2A702454'.toLowerCase()) &&
+      swapCalls[0].calldata.toLowerCase().includes('bb4CdB9CBd36B01bD1cBaEBF2De08d9173bc095c'.toLowerCase())
```
Comment (chef-jojo) [inline apps/web/src/views/Swap/V3Swap/hooks/useWallchain.ts:98]: how about use `extractTokensFromTrade` to check src/dist matching?
move the address into config `apps/web/src/config/wallchain.ts`

File: apps/web/src/views/Swap/V3Swap/hooks/useWallchain.ts [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+              amountIn,
+              isPermit: false,
+            },
+            masterInput,
+            witness,
+          )
+
+          return {
+            address: data.to as `0x${string}`,
+            calldata: data.data as `0x${string}`,
```
Comment (chef-jojo) [inline apps/web/src/views/Swap/V3Swap/hooks/useWallchain.ts:188]: can we add a checking the `data.data` match the `previousSwapCalls[0].data` here?

File: apps/web/src/views/Swap/V3Swap/hooks/useWallchain.ts [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+
+  useEffect(() => {
+    if (!walletClient) {
+      return
+    }
+    const sdk = new WallchainSDK({
+      keys: WallchainKeys,
+      provider: walletClient.transport as TOptions['provider'],
+      overrideAddresses,
+    })
```
Comment (chef-jojo) [inline apps/web/src/views/Swap/V3Swap/hooks/useWallchain.ts:93]: Let's move the sdk initialisation inside the condition, would be great to use lazy import `await import('@wallchain/sdk')`

File: apps/web/src/views/Swap/V3Swap/containers/SwapCommitButton.tsx [hunk 4]
```diff
@@ -83,7 +84,14 @@ export const SwapCommitButton = memo(function SwapCommitButton({
...
   const [isRoutingSettingChange, resetRoutingSetting] = useRoutingSettingChanged()
   const slippageAdjustedAmounts = useSlippageAdjustedAmounts(trade)
-  const routerAddress = SWAP_ROUTER_ADDRESSES[trade?.inputAmount?.currency?.chainId]
+
+  const deadline = useTransactionDeadline()
+  const [statusWallchain, approvalAddressForWallchain, wallchainMasterInput] = useWallchainApi(trade, deadline)
+
+  const routerAddress =
+    statusWallchain === 'found'
+      ? approvalAddressForWallchain
```
Comment (chef-jojo) [inline apps/web/src/views/Swap/V3Swap/containers/SwapCommitButton.tsx:93]: One more need to check, this approval address is dynamic from api. It'd be better hardcoded

File: apps/web/src/views/Swap/V3Swap/hooks/useWallchain.ts [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+  if (await sdk.supportsChain()) {
+    const resp = await sdk.checkForMEV({
+      from: account,
+      to: swapCalls[0].address,
+      value: swapCalls[0].value,
+      data: swapCalls[0].calldata,
+    })
+
+    if (resp.MEVFound) {
+      const approvalFor = await sdk.getSpenderForAllowance()
```
Comment (chef-jojo) [inline apps/web/src/views/Swap/V3Swap/hooks/useWallchain.ts:54]: let's simply hardcode the address using overrideAddreses above

**Pull request closed without merging**